### PR TITLE
Rules CRUD

### PR DESCRIPTION
**Pull request by jillcary**
# Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

# Checklist:

- [x] The code follows the code style and conventions of this project.
- [x] The change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] All new and existing tests passed.

# Description of the change

This change includes the CRUD operations for rules management using the StorageAdapter service. 
* Create Rule
* Update Rule
* Delete Rule
* Get list of rules with parameters (order, limit, skip, groupId)
* Get single Rule by id

This change also includes:
* Remove the "Action" on the rule class as it is not needed for MVP. 
* Remove ConditionsList has also been removed as it is not needed. 
* Logging when a rules request returns null

### CODE REVIEW

File: app/com/microsoft/azure/iotsolutions/devicetelemetry/services/Rules.java [hunk 1]
```diff
@@ -2,161 +2,290 @@
...
+                }
+
+                if (result.getStatus() == NOT_FOUND) {
+                    return null;
+                }
+
+                try {
+                    return getRuleServiceModelFromJson(Json.parse(result.getBody()));
+                } catch (Exception e) {
+                    log.error("Could not parse result from Key Value Storage");
```
Comment (dluc) [inline app/com/microsoft/azure/iotsolutions/devicetelemetry/services/Rules.java]: nit: log also the exception msg

File: app/com/microsoft/azure/iotsolutions/devicetelemetry/services/Rules.java [hunk 2]
```diff
@@ -2,161 +2,290 @@
...
+                    if (order.equalsIgnoreCase("asc")) {
+                        Collections.sort(ruleList);
+                    } else {
+                        Collections.sort(ruleList, Collections.reverseOrder());
+                    }
+
+                    // Check bounds on parameters
+                    if (skip >= ruleList.size()) {
+                        log.warn("Skip value greater than size of list");
+                        return null;
```
Comment (dluc) [inline app/com/microsoft/azure/iotsolutions/devicetelemetry/services/Rules.java]: nit: I would return an empty array

File: app/com/microsoft/azure/iotsolutions/devicetelemetry/services/Rules.java [hunk 3]
```diff
@@ -2,161 +2,290 @@
...
+
+                    if (order.equalsIgnoreCase("asc")) {
+                        Collections.sort(ruleList);
+                    } else {
+                        Collections.sort(ruleList, Collections.reverseOrder());
+                    }
+
+                    // Check bounds on parameters
+                    if (skip >= ruleList.size()) {
+                        log.warn("Skip value greater than size of list");
```
Comment (dluc) [inline app/com/microsoft/azure/iotsolutions/devicetelemetry/services/Rules.java]: nit: decrease to "Debug" level

File: app/com/microsoft/azure/iotsolutions/devicetelemetry/services/Rules.java [hunk 4]
```diff
@@ -2,161 +2,290 @@
...
+                        return null;
+                    } else if (skip < 0 || limit <= 0) {
```
Comment (dluc) [inline app/com/microsoft/azure/iotsolutions/devicetelemetry/services/Rules.java]: input validation should be done before calling external resoruces

File: app/com/microsoft/azure/iotsolutions/devicetelemetry/services/Rules.java [hunk 5]
```diff
@@ -2,161 +2,290 @@
...
+                        log.error("Key value storage parameter bounds error");
+                        throw new CompletionException(
+                            new InvalidInputException("Parameter bounds error"));
+                    } else if ((limit + skip) > ruleList.size()) {
+                        return ruleList.subList(skip, ruleList.size() - skip - 1);
+                    }
+
+                    return ruleList.subList(skip, limit);
```
Comment (dluc) [inline app/com/microsoft/azure/iotsolutions/devicetelemetry/services/Rules.java]: shouldn't this be `subList(skip, skip + limit - 1);` ?

File: app/com/microsoft/azure/iotsolutions/devicetelemetry/services/Rules.java [hunk 6]
```diff
@@ -2,161 +2,290 @@
...
+
+                } catch (Exception e) {
+                    log.error("Could not parse result from Key Value Storage");
```
Comment (dluc) [inline app/com/microsoft/azure/iotsolutions/devicetelemetry/services/Rules.java]: log exception too

File: app/com/microsoft/azure/iotsolutions/devicetelemetry/services/Rules.java [hunk 7]
```diff
@@ -2,161 +2,290 @@
...
+                    log.error("Key value storage request error: {}",
+                        error.getMessage());
+                    throw new CompletionException(
+                        new ExternalDependencyException(error.getMessage()));
+                }
+
+                try {
+                    return getRuleServiceModelFromJson(
+                        Json.parse(result.getBody()));
+                } catch (Exception e) {
```
Comment (dluc) [inline app/com/microsoft/azure/iotsolutions/devicetelemetry/services/Rules.java]: the exception should be rethrown and cause a 500 no?

File: app/com/microsoft/azure/iotsolutions/devicetelemetry/services/Rules.java [hunk 8]
```diff
@@ -2,161 +2,290 @@
...
+
+                        log.error("Key value storage ETag mismatch");
+                        throw new CompletionException(
+                            new ResourceOutOfDateException());
+                    }
+
+                    return getRuleServiceModelFromJson(
+                        Json.parse(result.getBody()));
+                } catch (Exception e) {
+
```
Comment (dluc) [inline app/com/microsoft/azure/iotsolutions/devicetelemetry/services/Rules.java]: ditto

File: app/com/microsoft/azure/iotsolutions/devicetelemetry/services/Rules.java [hunk 9]
```diff
@@ -2,161 +2,290 @@
...
-
-        return sampleRules;
+    private RuleServiceModel getRuleServiceModelFromJson(JsonNode response) {
+        ObjectMapper mapper = new ObjectMapper();
+        JsonNode jsonResultRule;
+
+        try {
+            jsonResultRule = mapper.readTree(response.findValue("Data").asText());
+        } catch (Exception e) {
+            log.error("Could not parse data from Key Value Storage");
```
Comment (dluc) [inline app/com/microsoft/azure/iotsolutions/devicetelemetry/services/Rules.java]: nit: log the JSON text

File: app/com/microsoft/azure/iotsolutions/devicetelemetry/webservice/v1/controllers/RulesController.java [hunk 10]
```diff
@@ -29,15 +41,46 @@ public RulesController(final IRules rules) {
...
+            order,
+            skip,
+            limit,
+            groupId)
+            .thenApply(ruleServiceModelList -> {
+                if (ruleServiceModelList != null &&
+                    ruleServiceModelList.size() > 0) {
+                    return ok(toJson(new RuleListApiModel(ruleServiceModelList)));
+                } else {
+                    return notFound();
```
Comment (dluc) [inline app/com/microsoft/azure/iotsolutions/devicetelemetry/webservice/v1/controllers/RulesController.java]: when the result is empty the service should return 200 OK and an empty list. 404 would be returned if the endpoint didn't exist.

File: app/com/microsoft/azure/iotsolutions/devicetelemetry/webservice/v1/controllers/RulesController.java [hunk 11]
```diff
@@ -29,15 +41,46 @@ public RulesController(final IRules rules) {
...
+    public CompletionStage<Result> get(String id) {
+
+        return
+            rules.getAsync(id)
+                .thenApply(
+                    rule -> {
+                        if (rule != null) {
+                            return ok(toJson(new RuleApiModel(rule)));
+                        } else {
+                            return notFound();
```
Comment (dluc) [inline app/com/microsoft/azure/iotsolutions/devicetelemetry/webservice/v1/controllers/RulesController.java]: 👍  404 in this case is correct

File: app/com/microsoft/azure/iotsolutions/devicetelemetry/webservice/v1/controllers/RulesController.java [hunk 12]
```diff
@@ -46,25 +89,96 @@ public Result get(String id) {
...
      */
-    public Result post() {
-        return ok(toJson(new RuleApiModel(this.rules.post(new RuleServiceModel()))));
+    public CompletionStage<Result> post() {
+
+        JsonNode jsonBody = request().body().asJson();
+
+        if (jsonBody == null) {
+            log.warn("Error retrieving post request body");
+            return null;
```
Comment (dluc) [inline app/com/microsoft/azure/iotsolutions/devicetelemetry/webservice/v1/controllers/RulesController.java]: return a 400 Bad Request

File: app/com/microsoft/azure/iotsolutions/devicetelemetry/webservice/v1/controllers/RulesController.java [hunk 13]
```diff
@@ -46,25 +89,96 @@ public Result get(String id) {
...
+            jsonBody.findValue("Name").asText(),
+            jsonBody.findValue("Enabled").asBoolean(),
+            jsonBody.findValue("Description").asText(),
+            jsonBody.findValue("GroupId").asText(),
+            jsonBody.findValue("Severity").asText(),
+            conditions);
+
+        return rules.postAsync(ruleServiceModel)
+            .thenApply(newRule -> {
+                if (newRule != null) {
```
Comment (dluc) [inline app/com/microsoft/azure/iotsolutions/devicetelemetry/webservice/v1/controllers/RulesController.java]: postAsync should never return null. In case of failure it should throw an exception

File: app/com/microsoft/azure/iotsolutions/devicetelemetry/webservice/v1/controllers/RulesController.java [hunk 14]
```diff
@@ -46,25 +89,96 @@ public Result get(String id) {
...
      *
      * @return updated rule
      */
-    public Result put(String id) {
-        return ok(toJson(new RuleApiModel(this.rules.put(new RuleServiceModel()))));
+    public CompletionStage<Result> put(String id) {
+        JsonNode jsonBody = request().body().asJson();
+
+        if (jsonBody == null) {
+            log.warn("Error retrieving post request body");
```
Comment (dluc) [inline app/com/microsoft/azure/iotsolutions/devicetelemetry/webservice/v1/controllers/RulesController.java]: I would rephrase to "The request payload is empty"

File: app/com/microsoft/azure/iotsolutions/devicetelemetry/webservice/v1/controllers/RulesController.java [hunk 15]
```diff
@@ -46,25 +89,96 @@ public Result get(String id) {
...
+            return CompletableFuture.completedFuture(
+                badRequest("Invalid input"));
```
Comment (dluc) [inline app/com/microsoft/azure/iotsolutions/devicetelemetry/webservice/v1/controllers/RulesController.java]: nit: "The request is empty"

File: app/com/microsoft/azure/iotsolutions/devicetelemetry/webservice/v1/controllers/RulesController.java [hunk 16]
```diff
@@ -46,25 +89,96 @@ public Result get(String id) {
...
+                jsonBody.findValue("Name").asText(),
+                jsonBody.findValue("DateCreated").asText(),
+                DateTime.now().toString(),
+                jsonBody.findValue("Enabled").asBoolean(),
+                jsonBody.findValue("Description").asText(),
+                jsonBody.findValue("GroupId").asText(),
+                jsonBody.findValue("Severity").asText(),
+                conditions);
+
+            return rules.postAsync(ruleServiceModel)
```
Comment (dluc) [inline app/com/microsoft/azure/iotsolutions/devicetelemetry/webservice/v1/controllers/RulesController.java]: putAsync

File: app/com/microsoft/azure/iotsolutions/devicetelemetry/webservice/v1/controllers/RulesController.java [hunk 17]
```diff
@@ -46,25 +89,96 @@ public Result get(String id) {
...
+                .thenApply(newRule -> {
+                    if (newRule != null) {
```
Comment (dluc) [inline app/com/microsoft/azure/iotsolutions/devicetelemetry/webservice/v1/controllers/RulesController.java]: putAsync should never return null, but throw an exception in case of failure

File: app/com/microsoft/azure/iotsolutions/devicetelemetry/webservice/v1/controllers/RulesController.java [hunk 18]
```diff
@@ -2,18 +2,30 @@
...
 import play.mvc.Controller;
 import play.mvc.Result;
 
+import java.util.ArrayList;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+
 import static play.libs.Json.toJson;
 
 public class RulesController extends Controller {
```
Comment (dluc) [inline app/com/microsoft/azure/iotsolutions/devicetelemetry/webservice/v1/controllers/RulesController.java]: nit: can we rename the controllers method and append "Async" ? if other controllers have the same issue, we can do it separately in a separate PR

File: app/com/microsoft/azure/iotsolutions/devicetelemetry/services/Rules.java [hunk 19]
```diff
@@ -2,161 +2,290 @@
...
+                    throw new CompletionException(
+                        new ExternalDependencyException(error.getMessage()));
+                }
+
+                if (result.getStatus() == NOT_FOUND) {
+                    return null;
+                }
+
+                try {
+                    return getRuleServiceModelFromJson(Json.parse(result.getBody()));
```
Comment (dluc) [inline app/com/microsoft/azure/iotsolutions/devicetelemetry/services/Rules.java]: when creating a RuleServiceModel, the Etag property should be set using the value from result object

File: app/com/microsoft/azure/iotsolutions/devicetelemetry/services/models/RuleServiceModel.java [hunk 20]
```diff
@@ -3,23 +3,22 @@
...
-public final class RuleServiceModel {
+public final class RuleServiceModel implements Comparable<RuleServiceModel> {
 
     private final String eTag;
     private final String id;
     private final String name;
-    private final DateTime dateCreated;
-    private final DateTime dateModified;
+    private final String dateCreated;
+    private final String dateModified;
```
Comment (dluc) [inline app/com/microsoft/azure/iotsolutions/devicetelemetry/services/models/RuleServiceModel.java]: change both to DateTime class

File: app/com/microsoft/azure/iotsolutions/devicetelemetry/services/Rules.java (hunk 4 above)
Comment (jillcary) [inline app/com/microsoft/azure/iotsolutions/devicetelemetry/services/Rules.java]: Moved what I could before call to storage. The skip > list length check must be after the call to StorageAdapter, since list length depends on the result.

File: app/com/microsoft/azure/iotsolutions/devicetelemetry/services/Rules.java (hunk 5 above)
Comment (jillcary) [inline app/com/microsoft/azure/iotsolutions/devicetelemetry/services/Rules.java]: Fixed. It should be the "to limit" which is the end of the list (exclusive) to ruleList.size().

File: app/com/microsoft/azure/iotsolutions/devicetelemetry/webservice/v1/controllers/RulesController.java (hunk 18 above)
Comment (jillcary) [inline app/com/microsoft/azure/iotsolutions/devicetelemetry/webservice/v1/controllers/RulesController.java]: Done. I believe StatusController is the only other controller impacted.

File: app/com/microsoft/azure/iotsolutions/devicetelemetry/services/Rules.java (hunk 7 above)
Comment (jillcary) [inline app/com/microsoft/azure/iotsolutions/devicetelemetry/services/Rules.java]: done.

File: app/com/microsoft/azure/iotsolutions/devicetelemetry/services/models/RuleServiceModel.java (hunk 20 above)
Comment (jillcary) [inline app/com/microsoft/azure/iotsolutions/devicetelemetry/services/models/RuleServiceModel.java]: Opted to keep as a string for serialization. See. issue for more info: https://github.com/Azure/device-telemetry-java/issues/57

File: app/com/microsoft/azure/iotsolutions/devicetelemetry/webservice/v1/controllers/RulesControllerAsync.java [hunk 21]
```diff
@@ -0,0 +1,186 @@
...
+import play.mvc.Result;
+
+import javax.print.DocFlavor;
+import java.util.ArrayList;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+
+import static play.libs.Json.toJson;
+
+public class RulesControllerAsync extends Controller {
```
Comment (dluc) [inline app/com/microsoft/azure/iotsolutions/devicetelemetry/webservice/v1/controllers/RulesControllerAsync.java]: I think you meant to rename the methods, not the class, right?

File: app/com/microsoft/azure/iotsolutions/devicetelemetry/webservice/v1/controllers/RulesControllerAsync.java [hunk 22]
```diff
@@ -0,0 +1,186 @@
...
+        return this.rules.getListAsync(
+            order,
+            skip,
+            limit,
+            groupId)
+            .thenApply(ruleServiceModelList -> {
+                if (ruleServiceModelList != null) {
+                    return ok(toJson(new RuleListApiModel(ruleServiceModelList)));
+                } else {
+                    return notFound();
```
Comment (dluc) [inline app/com/microsoft/azure/iotsolutions/devicetelemetry/webservice/v1/controllers/RulesControllerAsync.java]: should return OK with an empty list

File: app/com/microsoft/azure/iotsolutions/devicetelemetry/webservice/v1/controllers/RulesControllerAsync.java [hunk 23]
```diff
@@ -0,0 +1,186 @@
...
+                .thenApply(newRule -> ok(toJson(new RuleApiModel(newRule))))
+                .exceptionally(e -> {
+                    if (e.getCause() instanceof ResourceOutOfDateException) {
+                        return status(CONFLICT, Json.toJson(ruleServiceModel));
+                    } else {
+                        return internalServerError(e.getMessage());
+                    }
+                });
+
+        } catch (Exception e) {
```
Comment (dluc) [inline app/com/microsoft/azure/iotsolutions/devicetelemetry/webservice/v1/controllers/RulesControllerAsync.java]: if I understand the code correctly, the execution gets here if there is an error in lines 148-158. If that's the case, could you move the `rules.putAsync` after the try-catch?

File: app/com/microsoft/azure/iotsolutions/devicetelemetry/webservice/v1/controllers/RulesController.java [hunk 24]
```diff
@@ -34,14 +37,14 @@ public RulesController(final IRules rules) {
     }
 
     /**
-     * Return a list of rules. The list of rules can be paginated, and
-     * filtered by device and status. The list is sorted
+     * Return a listAsync of rules. The listAsync of rules can be paginated, and
```
Comment (dluc) [inline app/com/microsoft/azure/iotsolutions/devicetelemetry/webservice/v1/controllers/RulesController.java]: nit: revert this

File: app/com/microsoft/azure/iotsolutions/devicetelemetry/webservice/v1/controllers/RulesController.java (hunk 10 above)
Comment (dluc) [inline app/com/microsoft/azure/iotsolutions/devicetelemetry/webservice/v1/controllers/RulesController.java]: ping...

File: app/com/microsoft/azure/iotsolutions/devicetelemetry/webservice/v1/controllers/RulesController.java [hunk 25]
```diff
@@ -105,36 +108,37 @@ public RulesController(final IRules rules) {
...
      *
      * @return updated rule
      */
-    public CompletionStage<Result> put(String id) {
+    public CompletionStage<Result> putAsync(String id) {
         JsonNode jsonBody = request().body().asJson();
 
         if (jsonBody == null) {
-            log.warn("Error retrieving post request body");
+            log.warn("The request is empty");
```
Comment (dluc) [inline app/com/microsoft/azure/iotsolutions/devicetelemetry/webservice/v1/controllers/RulesController.java]: nit: use this sentence also in postAsync

File: app/com/microsoft/azure/iotsolutions/devicetelemetry/webservice/v1/controllers/RulesController.java [hunk 26]
```diff
@@ -37,8 +37,8 @@ public RulesController(final IRules rules) {
     }
 
     /**
-     * Return a listAsync of rules. The listAsync of rules can be paginated, and
-     * filtered by device and status. The listAsync is sorted
+     * Return a listof rules. The list of rules can be paginated, and
```
Comment (dluc) [inline app/com/microsoft/azure/iotsolutions/devicetelemetry/webservice/v1/controllers/RulesController.java]: typo "listof"
